src/validate.rs

```diff
@@ -1,23 +1,723 @@
 //! DNSSEC validation.
 //!
 //! **This module is experimental and likely to change significantly.**
-#![cfg(feature = "validate")]
-#![cfg_attr(docsrs, doc(cfg(feature = "validate")))]
+#![cfg(feature = "unstable-validate")]
+#![cfg_attr(docsrs, doc(cfg(feature = "unstable-validate")))]
 
 use crate::base::cmp::CanonicalOrd;
-use crate::base::iana::{DigestAlg, SecAlg};
+use crate::base::iana::{Class, DigestAlg, SecAlg};
 use crate::base::name::Name;
 use crate::base::name::ToName;
 use crate::base::rdata::{ComposeRecordData, RecordData};
 use crate::base::record::Record;
+use crate::base::scan::{IterScanner, Scanner};
 use crate::base::wire::{Compose, Composer};
-use crate::rdata::{Dnskey, Rrsig};
+use crate::base::Rtype;
+use crate::rdata::{Dnskey, Ds, Rrsig};
 use bytes::Bytes;
 use octseq::builder::with_infallible;
+use octseq::{EmptyBuilder, FromBuilder};
 use ring::{digest, signature};
+use std::boxed::Box;
 use std::vec::Vec;
 use std::{error, fmt};
 
+//----------- Key ------------------------------------------------------------
+
+/// A DNSSEC key for a particular zone.
+#[derive(Clone)]
+pub struct Key<Octs> {
+    /// The owner of the key.
+    owner: Name<Octs>,
+
+    /// The flags associated with the key.
+    ///
+    /// These flags are stored in the DNSKEY record.
+    flags: u16,
+
+    /// The raw public key.
+    ///
+    /// This identifies the key and can be used for signatures.
+    key: RawPublicKey,
+}
+
+//--- Construction
+
+impl<Octs> Key<Octs> {
+    /// Construct a new DNSSEC key manually.
+    pub fn new(owner: Name<Octs>, flags: u16, key: RawPublicKey) -> Self {
+        Self { owner, flags, key }
+    }
+}
+
+//--- Inspection
+
+impl<Octs> Key<Octs> {
+    /// The owner name attached to the key.
+    pub fn owner(&self) -> &Name<Octs> {
+        &self.owner
+    }
+
+    /// The flags attached to the key.
+    pub fn flags(&self) -> u16 {
+        self.flags
+    }
+
+    /// The raw public key.
+    pub fn raw_public_key(&self) -> &RawPublicKey {
+        &self.key
+    }
+
+    /// The signing algorithm used.
+    pub fn algorithm(&self) -> SecAlg {
+        self.key.algorithm()
+    }
+
+    /// Whether this is a zone signing key.
+    ///
+    /// From RFC 4034, section 2.1.1:
+    ///
+    /// > Bit 7 of the Flags field is the Zone Key flag.  If bit 7 has value
+    /// > 1, then the DNSKEY record holds a DNS zone key, and the DNSKEY RR's
+    /// > owner name MUST be the name of a zone.  If bit 7 has value 0, then
+    /// > the DNSKEY record holds some other type of DNS public key and MUST
+    /// > NOT be used to verify RRSIGs that cover RRsets.
+    pub fn is_zone_signing_key(&self) -> bool {
+        self.flags & (1 << 8) != 0
```

Comment:
This didn't correctly detect the ZSK flag for me when the input file contained a flags value of 256, for which RFC 4034 says _"Value 256 indicates that the Zone Key bit (bit 7) in the Flags field has value 1".

The following modified code worked for me:

```suggestion
        self.flags & 0b0000_0001_0000_0000 != 0
```